libass/ass.c

```diff
@@ -971,54 +965,122 @@ static int process_line(ASS_Track *track, char *str)
     return 0;
 }
 
-static int process_text(ASS_Track *track, char *str)
+static int process_text(ASS_Track *track, ASS_StringView *str)
 {
-    char *p = str;
     while (1) {
-        char *q;
         while (1) {
-            if ((*p == '\r') || (*p == '\n'))
-                ++p;
-            else if (p[0] == '\xef' && p[1] == '\xbb' && p[2] == '\xbf')
-                p += 3;         // U+FFFE (BOM)
+            if (str->len && (*str->str == '\r' || *str->str == '\n'))
+                ++str->str, --str->len;
+            else if (str->len >= 3 &&
+                     str->str[0] == '\xef' && str->str[1] == '\xbb' && str->str[2] == '\xbf')
+                str->str += 3, str->len -= 3; // U+FFFE (BOM)
             else
                 break;
         }
-        for (q = p; ((*q != '\0') && (*q != '\r') && (*q != '\n')); ++q) {
-        };
-        if (q == p)
+        ASS_StringView line = *str;
+        for (line.len = 0;
+             line.len < str->len && line.str[line.len] != '\r' && line.str[line.len] != '\n';
+             line.len++)
+             ;
+        if (!line.len || line.len == str->len)
             break;
-        if (*q != '\0')
-            *(q++) = '\0';
-        process_line(track, p);
-        if (*q == '\0')
-            break;
-        p = q;
+        process_line(track, line);
+        str->str += line.len, str->len -= line.len;
     }
     // there is no explicit end-of-font marker in ssa/ass
     if (track->parser_priv->fontname)
         decode_font(track);
     return 0;
 }
 
+static int process_text_full(ASS_Track *track, ASS_StringView str)
+{
+    int ret = process_text(track, &str);
+    if (ret < 0)
+        return ret;
+
+    if (!str.len)
+        return 0;
+
+    // Absurd? Yes.
+    if (str.len >= SIZE_MAX - 1)
+        return -ENOMEM;
+
+    char *buf = malloc(str.len + 1);
+    if (!buf)
+        return -ENOMEM;
+
+    str.str = buf;
```

Comment:
Buffer with garbage. Do you mean `str.str = ass_copy_string(str);`?

Reply:
Whoooooops! Almost; I'd meant to terminate the line with a `\n`. Fixed.